### PR TITLE
Clang cleanup

### PR DESCRIPTION
**Pull request by jjwilke**
This is an initial set of changes to clean up how the compiler wrapper works. This primarily gets rid of the following anti-pattern:

1. Pragma changes a statement using the rewriter
2. Later pragma analyzes the rewritten statement, but sees the original AST.

There was pretty terrible hackery around this issue to make things work. This is the first wave of changes needed to get rid of that. Pragmas now are allowed to only do 1 of 3 things:

1. Analyze a statement and record some info about it
2. Modify the AST data structure for the statement
3. Rewrite the statement. No further analyses or modifications allowed.

More cleanup is required - but we should merge this back ASAP.

### CODE REVIEW

File: bin/clang/annotatePragma.cc [hunk 1]
```diff
@@ -166,16 +166,15 @@ namespace annotate {
 namespace detail {
 
 annotate::ToolInfo
-getToolInfo(clang::SourceLocation Loc, clang::CompilerInstance &CI,
-            std::map<std::string, std::list<std::string>> &&PragmaStrings) {
+getToolInfo(clang::SourceLocation Loc, std::map<std::string, std::list<std::string>> &&PragmaStrings) {
```
Comment (calewis) [inline bin/clang/annotatePragma.cc]: We really need to get NM to let us run format on pushes or PRs

File: bin/clang/clangGlobals.h [hunk 2]
```diff
@@ -50,10 +57,116 @@ namespace clang {
...
+  SHADOWIZE_MODE = 3,
+  PUPPETIZE_MODE = 4,
+  NUM_MODES
+};
+
 }
+
+class SkeletonASTVisitor;
+class FirstPassASTVisitor;
+struct SSTPragma;
```
Comment (calewis) [inline bin/clang/clangGlobals.h]: struct vs class is annoying in a few places.  Also sst has a lot of mixed struct and class declarations which aren't errors, but aren't great.

File: bin/clang/clangGlobals.h [hunk 3]
```diff
@@ -50,10 +57,116 @@ namespace clang {
...
+class SSTReplacePragma;
+class SSTNullVariablePragma;
+struct PragmaConfig {
+  std::map<clang::Decl*,SSTNullVariablePragma*> nullVariables;
+  std::map<clang::FunctionDecl*,std::set<SSTPragma*>> functionPragmas;
+  std::map<const clang::DeclContext*,SSTNullVariablePragma*> nullSafeFunctions;
```
Comment (calewis) [inline bin/clang/clangGlobals.h]: Should we switch these to the llvm optimized structures?

File: sstmac/software/libraries/compute/compute_api.cc [hunk 4]
```diff
@@ -154,3 +154,4 @@ sstmac_compute_loop4(uint64_t isize, uint64_t jsize, uint64_t ksize, uint64_t ls
   sstmac::sw::OperatingSystem::currentThread()->parentApp()
     ->computeLoop(num_loops, nflops_per_loop, nintops_per_loop, bytes_per_loop);
 }
+
```
Comment (calewis) [inline sstmac/software/libraries/compute/compute_api.cc]: Remove white space changes?

File: tests/clang_src2src/pragma_sst_loop_count.cc [hunk 5]
```diff
@@ -11,10 +11,11 @@ int fxn()
   double* x = new double[10];
   int* idx = new int[5];
   int loopMax = 5;
+  int sstLoopCount = loopMax/2;
 #pragma sst compute
   for (int i=0; i < loopMax; ++i){
     mul *= i;
-    #pragma sst loop_count loopMax/2
+    #pragma sst loop_count sstLoopCount
```
Comment (calewis) [inline tests/clang_src2src/pragma_sst_loop_count.cc]: Why the change here?

Comment (jjwilke) [reply tests/clang_src2src/pragma_sst_loop_count.cc]: Loop count pragmas are much more limited in what they can be. They can only be a constant or variable name, not a whole expression.

File: bin/clang/clangGlobals.h (hunk 3 above)
Comment (jjwilke) [reply bin/clang/clangGlobals.h]: Not yet. Maybe after this PR.

File: bin/clang/clangGlobals.h (hunk 2 above)
Comment (jjwilke) [reply bin/clang/clangGlobals.h]: Agreed. Clean up later.

File: bin/clang/astVisitor.cc [hunk 6]
```diff
@@ -106,8 +121,9 @@ static std::string appendText(clang::Expr* expr, const std::string& toAppend)
 }
 
 void
-ASTVisitorCmdLine::setup()
+CompilerGlobals::setup(clang::CompilerInstance* CI)
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/astVisitor.cc]: Simplification sounds great. Issue?

File: bin/clang/annotatePragma.cc [hunk 7]
```diff
@@ -191,69 +190,54 @@ getToolInfo(clang::SourceLocation Loc, clang::CompilerInstance &CI,
...
-void SSTAnnotatePragmaImpl::activatePuppetize(clang::Stmt *S,
-                                              clang::Rewriter &R,
-                                              PragmaConfig &Cfg) {
+void SSTAnnotatePragmaImpl::activatePuppetize(clang::Stmt *S) {
   if (auto ParentFunc = getParentFunctionDecl(S, Ctx)) {
-    R.InsertTextBefore(getStart(ParentFunc),
-                       annotationStr(Ti_.Name(), Ti_.Args(), getLines(Sm, S)));
+    CompilerGlobals::rewriter.InsertTextBefore(getStart(ParentFunc),
+                       annotationStr(Ti_.Name(), Ti_.Args(), getLines(CompilerGlobals::SM(), S)));
     S->dumpColor();
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/annotatePragma.cc]: Dump predates the PR. Leave for later PR.

File: skeletons/hpcg-3.0/src/TestNorms.cpp [hunk 8]
```diff
@@ -32,7 +32,7 @@ int TestNorms(TestNormsData & testnorms_data) {
  double mean_delta = 0.0;
 #pragma sst compute
  for (int i= 0; i<testnorms_data.samples; ++i) mean_delta += (testnorms_data.values[i] - testnorms_data.values[0]);
-#pragma sst replace mean 1
+#pragma sst init 1
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply skeletons/hpcg-3.0/src/TestNorms.cpp]: Worth considering. The init pragma has always been either a DeclStmt init or BinaryOperator RHS.  I'd prefer to leave the name for now and address pragma names in a separate PR.

File: bin/clang/util.cc [hunk 9]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
...
+    case tok::string_literal: {
+      std::stringstream sstr;
+      getLiteralDataAsString(tok, sstr);
+      std::string varName = sstr.str();
+      return nameToExpr(ctx, varName, loc);
+    }
+    case tok::numeric_constant: {
+      std::stringstream sstr;
+      getLiteralDataAsString(tok, sstr);
+      std::string valueText = sstr.str();
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Done.

File: sstmac/software/libraries/compute/compute_api.cc (hunk 4 above)
Comment (jjwilke) [reply sstmac/software/libraries/compute/compute_api.cc]: Done.

File: tests/clang_src2src/pragma_sst_loop_count.cc (hunk 5 above)
Comment (jjwilke) [reply tests/clang_src2src/pragma_sst_loop_count.cc]: Although this limits somewhat flexibility in the skeleton, this allows us to directly modify the AST and greatly simplifies logic in the rewriter.

File: skeletons/lulesh2.0.3/lulesh.cc [hunk 10]
```diff
@@ -661,7 +661,7 @@ void CalcFBHourglassForceForElems( Domain &domain,
    Index_t numElem8 = numElem * 8 ;
 
    Real_t *fx_elem; 
-   Real_t *fy_elem; 
+   Real_t *fy_elem;
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply skeletons/lulesh2.0.3/lulesh.cc]: Reverted change.

File: bin/clang/util.cc [hunk 11]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
...
+      if (valueText.find(".") == std::string::npos){
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Done.

File: bin/clang/util.cc [hunk 12]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
...
+        return flit;
+      }
+      break;
+    }
+    default:
+      return nullptr;
+  }
+}
+
+void getLiteralDataAsString(const Token &tok, std::ostream &os)
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: May not be used, but function is useful and should be kept.

File: bin/clang/util.cc [hunk 13]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
   return uniqueFilePrefix;
 }
 
+static Expr* nameToExpr(DeclContext* ctx, const std::string& name, clang::SourceLocation loc)
+{
+  std::vector<NamedDecl*> matches;
+  for (auto* d : ctx->decls()){
+    if (isa<NamedDecl>(d)){
+      NamedDecl* nd = cast<NamedDecl>(d);
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: I'd prefer to just leave it as is for now.

File: bin/clang/util.cc [hunk 14]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
...
+      if (nd->getNameAsString() == name){
+        matches.push_back(nd);
+      }
+    }
+  }
+
+  if (!matches.empty()){
+    NamedDecl* nd = matches.back();
+    VarDecl* vd = cast<VarDecl>(nd);
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: added error check

File: bin/clang/util.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Changed. Done.

File: bin/clang/util.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Done.

File: bin/clang/util.cc [hunk 15]
```diff
@@ -206,3 +206,91 @@ makeCxxName(const std::string& name)
...
+      std::string varName = sstr.str();
+      return nameToExpr(ctx, varName, loc);
+    }
+    case tok::numeric_constant: {
+      std::stringstream sstr;
+      getLiteralDataAsString(tok, sstr);
+      std::string valueText = sstr.str();
+      if (valueText.find(".") == std::string::npos){
+        //make an integer literal
+        int i = std::atoi(valueText.c_str());
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Done.

File: bin/clang/util.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/util.cc]: Maybe not, but the function is still useful in the future.

File: bin/clang/replacePragma.cc [hunk 16]
```diff
@@ -45,177 +45,98 @@ Questions? Contact sst-macro-help@sandia.gov
...
-          cfg.replacedExprs.push_back(cfg.parents.back());
-          break;
-        default:
-          //any other use, means this variable is not "connected" to anything else
-          cfg.replacedExprs.push_back(expr);
-          break;
+  bool TraverseDeclStmt(DeclStmt* stmt, DataRecursionQueue* queue = nullptr){
+    const Decl* d = stmt->getSingleDecl();
+    if (d && d->getKind() == Decl::Var){
+      const VarDecl* vd = cast<const VarDecl>(d);
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/replacePragma.cc]: Will leave as is for now. `getKind` is faster.

File: bin/clang/replacePragma.cc [hunk 17]
```diff
@@ -45,177 +45,98 @@ Questions? Contact sst-macro-help@sandia.gov
...
-        break;
-      default:
-        if (PreVisit){
-          //break the connection to prev parent
-          cfg.parents.push_back(expr);
-        } else { //restore the original parent
-          cfg.parents.pop_back();
-        }
+  bool matchingDeclRef(Expr* expr){
+    auto* underlying = SkeletonASTVisitor::getUnderlyingExpr(expr);
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/replacePragma.cc]: I prefer `auto*` to indicate this should be a pointer type.

File: bin/clang/replacePragma.cc [hunk 18]
```diff
@@ -45,177 +45,98 @@ Questions? Contact sst-macro-help@sandia.gov
...
-  bool operator()(const CallExpr* expr, ExprRole role,
-                  CompilerInstance& CI,
-                  ReplacementConfig& cfg){
-    visit(expr, role, CI, cfg);
-    return false;
+  bool TraverseArraySubscriptExpr(ArraySubscriptExpr* expr, DataRecursionQueue* queue = nullptr){
+    if (matchingDeclRef(expr->getBase())){
+      replaced_.push_back(expr);
+      return true;
+    } else {
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/replacePragma.cc]: I think I prefer it as is?

File: bin/clang/replacePragma.cc [hunk 19]
```diff
@@ -45,177 +45,98 @@ Questions? Contact sst-macro-help@sandia.gov
...
+      return Parent::TraverseCallExpr(expr);
+    }
   }
 
-  bool operator()(const MemberExpr* expr, ExprRole role,
-                  CompilerInstance& CI, ReplacementConfig& cfg){
-    visit(expr, expr->getFoundDecl()->getNameAsString(), role, CI, cfg);
-    return false;
+  const std::list<const Expr*> replaced() const {
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/replacePragma.cc]: I think I actually prefer it as is?

File: bin/clang/astVisitor.cc [hunk 20]
```diff
@@ -2262,6 +2270,7 @@ SkeletonASTVisitor::TraverseCXXConstructorDecl(CXXConstructorDecl *D)
 
   IncrementGuard ig(insideCxxMethod_);
   PushGuard<CXXConstructorDecl*> pg(ctorContexts_, D);
+  PushGuard<FunctionDecl*> pgf(fxnContexts_, D);
   for (auto *I : D->inits()) {
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/astVisitor.cc]: I like having the `auto*` to indicate pointer type.

File: bin/clang/astVisitor.cc [hunk 21]
```diff
@@ -2565,6 +2576,13 @@ SkeletonASTVisitor::TraverseWhileStmt(WhileStmt* S, DataRecursionQueue*  /*queue
 bool
 SkeletonASTVisitor::TraverseForStmt(ForStmt *S, DataRecursionQueue*  /*queue*/)
 {
+  if (fxnContexts_.empty()){
+    errorAbort(S, *ci_, "Traversing ForStmt with no function context");
+  }
+
+  FunctionDecl* fd = fxnContexts_.back();
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/astVisitor.cc]: Perhaps? If you don't use it, not necessary.

File: bin/clang/astVisitor.h [hunk 22]
```diff
@@ -54,6 +54,95 @@ Questions? Contact sst-macro-help@sandia.gov
...
+
+/**
+ * Exception-safe pushing back on a list
+ * Forces clean up even if exceptions get thrown
+ */
+template <class T>
+struct PushGuard {
+  template <class U>
+  PushGuard(std::list<T>& theList, U&& t) : myList(theList) {
+    myList.push_back(t);
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/astVisitor.h]: Yes. Fixed.

File: bin/clang/astVisitor.h [hunk 23]
```diff
@@ -54,6 +54,95 @@ Questions? Contact sst-macro-help@sandia.gov
 #define visitFxn(cls) \
   bool Visit##cls(clang::cls* c){ return TestStmtMacro(c); }
 
+static constexpr int IndexResetter = -1;
+
+/**
+ * Exception-safe pushing back on a list
+ * Forces clean up even if exceptions get thrown
+ */
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/astVisitor.h]: There's a `pop_back` in the destructor. Added some clarifying comments.

File: bin/clang/computePragma.cc [hunk 24]
```diff
@@ -47,20 +47,152 @@ Questions? Contact sst-macro-help@sandia.gov
...
+{
+  if (tokens.size() != 1){
+    errorAbort(loc, CI, "loop_count pragma should taken a single token as the pragma argument");
+  }
+  loopCountToken_ = tokens.front();
+  loopCount_ = getSingleString(tokens,CI);
+}
+
+void
+SSTLoopCountPragma::transformWhileLoop(Stmt* s, Rewriter& r, PragmaConfig& cfg)
```
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/computePragma.cc]: I tried and got it really messy. Maybe there is, but it didn't seem worth it when I tried.

File: bin/clang/util.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (calewis) [reply bin/clang/util.cc]: But this is almost exactly equivalent to what dyn_cast does, so it not actually slower.

File: bin/clang/replacePragma.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (calewis) [reply bin/clang/replacePragma.cc]: Do we want to enforce that or developer preference?

File: bin/clang/replacePragma.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (calewis) [reply bin/clang/replacePragma.cc]: Okay, I prefer the alternative, but we can leave it.

File: bin/clang/replacePragma.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (jjwilke) [reply bin/clang/replacePragma.cc]: Developer preference (for now).

File: bin/clang/replacePragma.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (calewis) [reply bin/clang/replacePragma.cc]: Fixed in next PR
